llvm/unittests/Support/CharSetTest.cpp

```diff
@@ -0,0 +1,232 @@
+  StringRef Src(HelloE);
+  SmallString<64> Dst;
+
+  ErrorOr<EncodingConverter> Conv =
+      EncodingConverter::create(TextEncoding::IBM1047, TextEncoding::UTF8);
+
+  // Stop test if conversion is not supported.
+  if (!Conv) {
+    ASSERT_EQ(Conv.getError(),
+              std::make_error_code(std::errc::invalid_argument));
+    return;
+  }
+
+  std::error_code EC = Conv->convert(Src, Dst);
+
+  EXPECT_TRUE(!EC);
+  EXPECT_STREQ(HelloA, static_cast<std::string>(Dst).c_str());
+  Dst.clear();
+
+  // ABC string.
+  Src = ABCStrE;
+  EC = Conv->convert(Src, Dst);
+  EXPECT_TRUE(!EC);
+  EXPECT_STREQ(ABCStrA, static_cast<std::string>(Dst).c_str());
+  Dst.clear();
+
+  // Accent string.
+  Src = AccentE;
+  EC = Conv->convert(Src, Dst);
+  EXPECT_TRUE(!EC);
+  EXPECT_STREQ(AccentUTF, static_cast<std::string>(Dst).c_str());
+}
+
+TEST(CharSet, RoundTrip) {
+  ErrorOr<EncodingConverter> ConvToUTF16 =
+      EncodingConverter::create("IBM-1047", "UTF-16");
+  // Stop test if conversion is not supported (no underlying iconv support).
+  if (!ConvToUTF16) {
+    ASSERT_EQ(ConvToUTF16.getError(),
+              std::make_error_code(std::errc::invalid_argument));
+    return;
+  }
+  ErrorOr<EncodingConverter> ConvToUTF32 =
+      EncodingConverter::create("UTF-16", "UTF-32");
+  // Stop test if conversion is not supported (no underlying iconv support).
+  if (!ConvToUTF32) {
+    ASSERT_EQ(ConvToUTF32.getError(),
+              std::make_error_code(std::errc::invalid_argument));
+    return;
+  }
+  ErrorOr<EncodingConverter> ConvToEBCDIC =
+      EncodingConverter::create("UTF-32", "IBM-1047");
+  // Stop test if conversion is not supported (no underlying iconv support).
+  if (!ConvToEBCDIC) {
+    ASSERT_EQ(ConvToEBCDIC.getError(),
+              std::make_error_code(std::errc::invalid_argument));
+    return;
+  }
+
+  // Setup source string.
+  char SrcStr[256];
+  for (size_t I = 0; I < 256; ++I)
+    SrcStr[I] = (I + 1) % 256;
+
+  SmallString<99> Dst1Str, Dst2Str, Dst3Str;
+
+  std::error_code EC = ConvToUTF16->convert(StringRef(SrcStr), Dst1Str);
+  EXPECT_TRUE(!EC);
+  EC = ConvToUTF32->convert(Dst1Str, Dst2Str);
+  EXPECT_TRUE(!EC);
+  EC = ConvToEBCDIC->convert(Dst2Str, Dst3Str);
+  EXPECT_TRUE(!EC);
+  EXPECT_STREQ(SrcStr, static_cast<std::string>(Dst3Str).c_str());
+}
+
+TEST(CharSet, ShiftState2022) {
+  // Earth string.
+  StringRef Src(EarthUTF);
+  SmallString<8> Dst;
+
+  ErrorOr<EncodingConverter> ConvTo2022 =
+      EncodingConverter::create("UTF-8", "ISO-2022-JP");
+  // Stop test if conversion is not supported (no underlying iconv support).
+  if (!ConvTo2022) {
+    ASSERT_EQ(ConvTo2022.getError(),
+              std::make_error_code(std::errc::invalid_argument));
+    return;
+  }
+
+  // Check that the string is properly converted.
+  std::error_code EC = ConvTo2022->convert(Src, Dst);
+  EXPECT_TRUE(!EC);
+  EXPECT_STREQ(EarthISO2022, static_cast<std::string>(Dst).c_str());
+}
+
+TEST(CharSet, ShiftState2022Partial) {
```

Comment:
Please rename this to better reflect what it is testing, which is a failure associated with the input encoding.
Please also add a test for a failure associated with the output encoding (e.g., input has a character that is not representable in the output character set).

Reply:
Thanks, I will rename. For output encoding failures, we have this existing testcase
```
  // Cyrillic string. Results in error because not representable in 1047.
  Src = CyrillicUTF;
  EC = Conv->convert(Src, Dst);
  EXPECT_EQ(EC, std::errc::illegal_byte_sequence);
```

Reply:
I should clarify: Please add a test for such a failure where the conversion is not backed by the internal table-based converter.

Reply:
I'll make a test from Cyrllic UTF-16 to IBM-1047, thanks!